crates/walrus-service/src/client/cli/args.rs

```diff
@@ -1015,6 +1201,89 @@ impl std::fmt::Display for BlobIdentity {
     }
 }
 
+/// Query for quilt patches by identifier.
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+pub struct QuiltPatchByIdentifier {
+    /// The quilt ID, which is the BlobID of the quilt.
+    pub quilt_id: BlobId,
+    /// The identifiers to read from the quilt.
+    pub identifiers: Vec<String>,
+}
+
+/// Query for quilt patches by tag.
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+pub struct QuiltPatchByTag {
+    /// The quilt ID, which is the BlobID of the quilt.
+    pub quilt_id: BlobId,
+    /// The tag key.
+    pub tag: String,
+    /// The tag value.
+    pub value: String,
+}
+
+/// Query for quilt patches by patch ID.
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+pub struct QuiltPatchByPatchId {
+    /// The quilt patch IDs.
+    pub quilt_patch_ids: Vec<QuiltPatchId>,
+}
+
+/// Selector for quilt patches.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum QuiltPatchSelector {
+    /// Patches by quilt_id and identifier.
+    ByIdentifier(QuiltPatchByIdentifier),
+    /// Patches by quilt_id and tag.
+    ByTag(QuiltPatchByTag),
+    /// Patches by quilt_patch_id.
+    ByPatchId(QuiltPatchByPatchId),
+    /// Read 'em all.
+    All(BlobId),
+}
+
+impl QuiltPatchSelector {
+    /// Get a selector from the command line arguments.
+    pub fn get_selector_from_args(
+        quilt_id: Option<BlobId>,
+        identifiers: Vec<String>,
+        tag: Option<String>,
+        value: Option<String>,
+        quilt_patch_ids: Vec<QuiltPatchId>,
+    ) -> Result<QuiltPatchSelector> {
+        if !identifiers.is_empty() {
+            Ok(QuiltPatchSelector::ByIdentifier(QuiltPatchByIdentifier {
+                quilt_id: quilt_id.expect("quilt_id should be present"),
+                identifiers: identifiers.clone(),
+            }))
+        } else if tag.is_some() && value.is_some() {
+            Ok(QuiltPatchSelector::ByTag(QuiltPatchByTag {
+                quilt_id: quilt_id.expect("quilt_id should be present"),
+                tag: tag.as_ref().expect("tag should be present").clone(),
+                value: value.as_ref().expect("value should be present").clone(),
+            }))
+        } else if !quilt_patch_ids.is_empty() {
+            Ok(QuiltPatchSelector::ByPatchId(QuiltPatchByPatchId {
+                quilt_patch_ids: quilt_patch_ids.clone(),
+            }))
```

Comment:
Nit: feel all the as_ref() and clone() are not necessary.

Question: are the command line parser configured in a way that the expect() should not be triggered? Just wonder if should instead returning Err() here to not crash the process (probably not a big deal right now)